Use C99 for loop to reduce the scope of the loop iterator

MisterDA opened the pull request:
C99 allows declaring the for loop variable iterator inside the for, reducing the scope of the variable. I think this improves readability and helps reasoning about the code. Now, if a for loop doesn't define its iterator, it's likely that the final value is needed out of the loop.

The diff was partially generated with variations of the following [Coccinelle](https://coccinelle.gitlabpages.inria.fr/website/) semantic patch (copyrighted by Richard Russon (flatcap) and released under the GPLv2):

```cocci
// Quite often, the loop variable in a 'for' loop isn't used elsewhere.
// C99 allows us to reduce the scope to *just* the 'for' loop.

@@
type T;
identifier I;
statement S;
expression E1, E2, E3;
@@

- T I;
  ... when != I
- for (I = E1; E2; E3)
+ for (T I = E1; E2; E3)
    S
  ... when != I
```

invoked as `for file in **/*.{c,h{,.in}}; do spatch -j8 --in-place --keep-comments --sp-file scoped-for-variable.cocci "$file"; done`. The script doesn't support nested for loops, loops reusing the same iterator, multiple variables defined in the for loop, so manual editing was needed afterwards.

I acknowledge that this is a behemoth of a PR, and would have preferred to be able to generate the diff automatically. To ease reviewer's life, I've kept it focused: there's no shadowing of variables, I'm not touching complex for loops (where there might be iterators of different types initialized and used). I'm not changing the type of anything.

No change entry needed?

MisterDA committed:
Remove duplicated comment end

MisterDA committed:
Reduce the scope of for loop iterators

ghost commented:
I like the idea and the use of coccinelle to help trusting these changes.

However, I have a few concerns:
- some of the changes have not been done with the coccinelle snippet you mentioned (e.g. changes to the macros in `runtime/bigarray.c`. Because of this, I would like this changeset split into multiple commits, with one commit per coccinnelle semantic patch.
- I also would like to see the changes to macros in a separate commit (because they might trigger shadowing warnings on some platforms, so it would be easier to push 90+% of that diff and drop the troublesome commits).

gasche commented on testsuite/tests/frame-pointers/fp_backtrace.c:
```diff
@@ -108,10 +108,11 @@ static void print_symbol(const char* symbol, const regmatch_t* match)
   const char* symbol = NULL;
 
-  for (struct frame_info* fi = __builtin_frame_address(0); fi; fi = next) {
+  for (struct frame_info* fi = __builtin_frame_address(0), *next = NULL;
```
This is hard to read for me. I think you should either use `struct frame_info *fi = ..., *next = ...` or put `next` back as a global.

MisterDA replied:
> I think you should either use `struct frame_info *fi = ..., *next = ...`

You mean in the scope above the loop, right?

We're doing clever things elsewhere… like traversing an array and counting the indices at the same time!

```c
value l;
intnat i;
/* ... */
for (i = 0, l = al; l != Val_emptylist; l = Field(l, 1), i++) {
  arrays[i] = Field(l, 0);
  offsets[i] = 0;
  lengths[i] = caml_array_length(Field(l, 0));
}
```

I'll try to harmonize walking over lists.

gasche replied:
Iterating on two things in parallel is okay, my comment was specifically about the use of `type* foo` for the first argument and `type *bar` for the second.

> I'll try to harmonize walking over lists.

Please don't mix too many changes in the current PR, every non-automatic change you add makes it noticeably harder to review. If it was just automatic changes it would be easier, manual changes need a different review approach and they don't mix well.

gasche commented on runtime/printexc.c:
```diff
@@ -95,10 +95,10 @@ CAMLexport char * caml_format_exception(value exn)
-  i = buf.ptr - buf.data + 1;
-  res = caml_stat_alloc_noexc(i);
+  len = buf.ptr - buf.data + 1;
+  res = caml_stat_alloc_noexc(len);
```
This part of the diff was edited manually. I reread it and it looks fine.

gasche reviewed (commented):
It is a change that touches all C files for a small benefit, but I must say that I like it, I do find the resulting code more readable. I also like the idea that we are using Coccinelle to work on the OCaml C runtime, best-in-class tools for best-in-class code :-)

I don't like the changes to the `yacc/` subdirectory. That code was written at a different time in a different style, and I find that the change hurt the consistency of the resulting code, in many places making the result less pleasant to read -- for example we started with global iteration variables `int i, j, k;` and now we only have `int k;`, the naming becomes harder to understand. I think you should get rid of the yacc/ changes.

MisterDA committed:
Introduce a new variable to avoid re-using a for loop iterator

MisterDA committed:
Change macros to reduce the scope of the for loop iterator

MisterDA committed:
Reduce the scope of the for loop iterator, avoid a shadowing warning

The function `caml_alloc_9` defines an `i` parameter. Compilers may
warn that defining local variables named `i` shadow this definition.
Pick the next available letter for the iterator.

MisterDA commented:
Thanks for the prompt reviews.

> `yacc/` subdirectory. [...] I think you should get rid of the `yacc/` changes.

Done, reverted.

> my comment was specifically about the use of `type* foo` for the first argument and `type *bar` for the second.

Indeed, I had missed the subtle difference. The new patch has the stars attached to the name. I used `sed`, I don't know how to express that in Coccinelle.

> - some of the changes have not been done with the Coccinelle snippet you mentioned (e.g. changes to the macros in `runtime/bigarray.c`. Because of this, I would like this changeset split into multiple commits, with one commit per Coccinelle semantic patch.

It's too difficult. The Coccinelle semantic patch has some issues, and there was some `sed` involved afterwards. If you insist on going that way, I'll probably withhold this PR for a while. In a ideal world I'd love to learn Coccinelle, but the manual is a strange beast.

> - I also would like to see the changes to macros in a separate commit (because they might trigger shadowing warnings on some platforms, so it would be easier to push 90+% of that diff and drop the troublesome commits).

Done.

gasche reviewed (approved):
I'm okay with the change -- thanks for splitting the commits, this is helpful.

This feels like a change that people could have opinions about, so I will wait a few days before going ahead with a merge.

yallop commented:
I haven't reviewed this in any detail, but I think it's a positive change.  It would be good (but not in this PR) to similarly modernize some other aspects of the C code in the runtime, e.g. using C99-style mixing of declarations and code to avoid declaring variables without initializing them.

dustanddreams reviewed (approved):
Thanks for splitting this in separate commits, this was much easier to review this way.

Looking good to me.

MisterDA commented:
I've added a changes entry (I had git hiccup in between).

NickBarnes reviewed (approved):
LGTM. I agree with @yallop that there are other moves in this general direction which I would also like.
One thing which this PR makes very clear is that there are a lot of different types (`int`, `intnat`, `uintnat`, `mlsize_t`, etc, etc) used for broadly the same purpose (indexing into an array); IWBNI there were some consistency (not on this PR, but in the future).

shindere commented:
Approving this on behalf of @NickBarnes
